newlib/libc/machine/arc64/strlen.S

```diff
@@ -0,0 +1,216 @@
+
+   Redistribution and use in source and binary forms, with or without
+   modification, are permitted provided that the following conditions are met:
+
+   1) Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+   2) Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+   3) Neither the name of the Synopsys, Inc., nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+   POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <sys/asm.h>
+
+; Code Brief:
+; Correct input pointer alignment and perform four 8 byte block search
+; This code is optimized for big strings, but also performs well for small
+; ones
+;
+; R0 const char* ptr (string to measure)
+; ret (R0):
+;		- unsigned (string size)
+;
+
+#if defined (__ARC64_ARCH64__)
+
+ENTRY (strlen)
+
+; If address is 4 byte aligned, we can just directly read 4 bytes
+bmsk.f	r3,		r0,		1
+beq.d			@.L_4B_step
+
+; Store r0 for size calculation when returning
+MOVP	r13,	r0
+
+subl	r3,		r3,		1		; [0]
+asl		r3,		r3,		2		; [1]
+
+; Jump depending on the alignment
+bi	    [r3]
+; Read 3 bytes
+ldb.ab	r10,	[r13,	+1]
+cmp		r10,	0
+beq 			@.L_return_size
+nop
+; Read 2 bytes
+ldb.ab	r10,	[r13,	+1]
+cmp		r10,	0
+beq				@.L_return_size
+nop
+; Read 1 byte
+ldb.ab	r10,	[r13,	+1]
+cmp		r10,	0
+beq				@.L_return_size
+
+; By this point, r13 is either 4 or 8 byte aligned
+.L_4B_step:
+
+; If aligned to 8 bytes, just jump ahead
+andl.f	0,		r13,	0b111
+beq.d	@.L_start_4_8B_search
+; Always reset bit mask that will encode NULL byte location
+xorl	r6,		r6,		r6
+
+ld.ab	r10,	[r13,	+4]
+sub		r11,	r10,	NULL_32DT_1
+bic		r11,	r11,	r10
+tst		r11,	NULL_32DT_2
+bne				@.L_reread_last_4bytes
+
+.L_start_4_8B_search:
+; Setup byte detector (more information bellow) [3] [5]
+movhl	r8,		NULL_32DT_1
+movhl	r9,		NULL_32DT_2
+
+orl		r8,		r8,		NULL_32DT_1
+orl		r9,		r9,		NULL_32DT_2
+
+ldl.ab	r2,		[r13,	+8]
+
+.L_4_8B_search:
+		ldl.ab	r3,		[r13,	+8]
```

Comment:
I would move this instruction too outside of the loop, such that I can take advantage of the `LDDL` instruction when it exists.